docs: add explanation about creating a slice

SamirPS opened the pull request:
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

sergiusens commented:
I am a bit confused. Shouldn't how to create slices be part of the chisel documentation?

cjdcordeiro commented:
> I am a bit confused. Shouldn't how to create slices be part of the chisel documentation?

It is indeed a fine line to walk. We were asked to have some Chisel documentation in the Rockcraft docs (which we do). On top of that, we've also received various support requests on building chiselled container images with Rockcraft. 

So, I don't think the Chisel **Reference** should be in Rockcraft, but we can definitely have some sections about it, as along as they are related with day-to-day work a typical Rockcrafter would need to do/understand, in order to build ROCKs.

cjdcordeiro commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,107 @@
+Creating our slice
+------------------
+
+Now we will create our openssl slice, but first, you need to clone this repository (our slice database):
```
```suggestion
You now have everything needed to actually define the OpenSSL slice that will install the content you're looking to have in the final ROCK.

Since you're looking to install just the OpenSSL binaries from the ``openssl`` package, what about naming this new slice _**bins**_? Let's go for it:

1. **What is the name of your slice definitions file?** It is a YAML file called ``openssl.yaml``
2. **What package name should be defined inside this file?** The package name is ``openssl``
3. **What is your slice name?** It should be called ``bins``
4. **What contents do you need from the OpenSSL package?** Just the binaries - ``/usr/bin/c_rehash`` and ``/usr/bin/openssl``
5. **Does your slice depend on any other package slice?** Yes, OpenSSL depends on ``libc6`` and ``libssl3``
    - **Do these two packages have slice definitions files upstream?** Yes, there's already a slice definitions file for [libc6](https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libc6.yaml) and another for [libssl3](https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libssl3.yaml)
        - If these dependencies were not present in the upstream Chisel release, you'd also need to create their corresponding slice definitions
    - **Which slices do you depend on then?** Since you only want the OpenSSL binaries, you might only need the libraries from ``libc6`` and ``libssl3``, as well as the configuration files from ``libc6`` and ``openssl`` itself.
    
With all this figured out, you can now create a new YAML file named ``openssl.yaml``, with the following content:

.. code-block:: yaml
    package: openssl
    slices:
        bins:
            essential:
            - libc6_libs
            - libc6_config
            - libssl3_libs
            - openssl_config
            contents:
              /usr/bin/c_rehash:
              /usr/bin/openssl:
            
        config:
            contents:
              /etc/ssl/private/:
              /etc/ssl/openssl.cnf:
              /usr/lib/ssl/certs:
              /usr/lib/ssl/openssl.cnf:
              /usr/lib/ssl/private:

Notice the unforeseen new slice ``config``. Because your OpenSSL binaries depend on the OpenSSL configuration files, and those weren't yet present anywhere in the Chisel releases upstream, you also need to create that slice! You may also ask _"why not put those configuration files inside the ``bins`` slice"_? You could! But we recommend, as a best practice, to separate and group contents according to their nature, as you may tomorrow need to create a new slice definition that only needs the OpenSSL configurations and not the binaries.

---

And that's it. This is your brand new slice definitions file, which will allow Chisel to install **just** the OpenSSL binaries (and their dependencies) into your ROCK! To learn about how to actually use this new slice definition file and publish it upstream for others to use, please check the following guides.
        
```

cjdcordeiro commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,107 @@
+
+Now we will create our openssl slice, but first, you need to clone this repository (our slice database):
+
+.. code-block:: sh
```
can you pls remove everything from here until the end of the file? the content is already covered in the suggested commit above

cjdcordeiro commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,117 @@
+    - if you find it, open it and try to find a slice name containing the bits and pieces you need from that package
+
+
+Structure of a slice definition file
```
```suggestion
Structure of a slice definitions file
```

cjdcordeiro commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,117 @@
+
+.. code-block:: yaml
+
+    # (req) Name of the package. The slice definition file should be named accordingly (eg. "openssl.yaml")
```
```suggestion
    # (req) Name of the package. 
    # The slice definition file should be named accordingly (eg. "openssl.yaml")
```

cjdcordeiro commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,117 @@
+                </path/to/mutable/file/with/default/text>: {text: FIXME, mutable: true}
+                </path/to/temporary/content>: {until: mutate}
+
+            # (opt) Mutation scripts, to allow for the reproduction of maintainer scripts, based on Starlark (https://github.com/canonical/starlark)
```
```suggestion
            # (opt) Mutation scripts, to allow for the reproduction of maintainer scripts,
            # based on Starlark (https://github.com/canonical/starlark)
```

cjdcordeiro commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,117 @@
+Find the dependencies of your package
+-------------------------------------
+
+So, the first step is to find the dependencies of the package for which you want to create a new slice definition (``openssl`` in this guide). You can easily do it by:
```
```suggestion
So, the first step is to find the dependencies of the package for which you want to create a new slice definitions (``openssl`` in this guide). You can easily do it by:
```

cjdcordeiro commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,117 @@
+
+    $ apt show openssl
+
+.. code-block::
```
```suggestion
.. code-block:: text
```

cjdcordeiro commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,117 @@
+    Bugs: https://bugs.launchpad.net/ubuntu/+filebug
+    Depends: libc6 (>= 2.34), libssl3 (>= 3.0.2-0ubuntu1.2)
+
+From the above output, you can confirm that ``openssl`` **depends on** ``libc6`` and ``libssl3``. So when creating your slice definition file for OpenSSL, you'll need to remember to include those packages slices as a dependency as well, whenever needed. Let's do that in the following section.
```
```suggestion
From the above output, you can confirm that ``openssl`` **depends on** ``libc6`` and ``libssl3``. So when creating your slice definitions file for OpenSSL, you'll need to remember to include those packages' slices as a dependency as well, whenever needed. Let's do that in the following section.
```

cjdcordeiro commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,117 @@
+----------------------------
+
+You now have everything needed to actually define the OpenSSL slice that will install the content you're looking to have in the final ROCK.
+Since you're looking to install just the OpenSSL binaries from the ``openssl`` package, what about naming this new slice **bins** ?  Let's go for it:
```
```suggestion
Since you're looking to install just the OpenSSL binaries from the ``openssl`` package, what about naming this new slice **bins**?  Let's go for it:
```

cjdcordeiro commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,117 @@
+
+.. code-block:: yaml
+
+    package: openssl
```
let's fix the indentation:

```
    package: openssl
    slices:
        bins:
            essential:
              - libc6_libs
              - libc6_config
              - libssl3_libs
              - openssl_config
            contents:
                /usr/bin/c_rehash:
                /usr/bin/openssl:

        config:
            contents:
                /etc/ssl/private/:
                /etc/ssl/openssl.cnf:
                /usr/lib/ssl/certs:
                /usr/lib/ssl/openssl.cnf:
                /usr/lib/ssl/private:
```

cjdcordeiro reviewed (changes requested):
some lists are not yet rendering well:

![image](https://user-images.githubusercontent.com/4047767/203742564-d670722d-4d4a-48f3-8748-844bca5df3b5.png)

---

![image](https://user-images.githubusercontent.com/4047767/203742805-5d09eb44-b7ac-4260-9b0e-3230dde40103.png)

SamirPS commented:
For the first list, it's because of the space between 2 and 3 ? and for the second, I don't see the problem

cjdcordeiro commented:
> For the first list, it's because of the space between 2 and 3 ? and for the second, I don't see the problem

it's the same spacing problem. these are 2 images. On the 2nd image, it's 4 and 5

SamirPS commented:
The spaces it's because of this: 
#. Scroll through the list of slice definition files, looking for your package name
    * if you find it, open it and try to find a slice name containing the bits and pieces you need from that package
if I remove the * the space disappear

cjdcordeiro commented:
y I know, but we want to keep the `*`

SamirPS commented:
> y I know, but we want to keep the `*.`

I try something to keep the *; tell me if you like it.

cjdcordeiro reviewed (changes requested):
sub-bullets are now rendering as quotes

cjdcordeiro reviewed (approved):
it renders well now

SamirPS commented:
Okay, thanks good to merge or another change to do?

cjdcordeiro commented:
did something change since the last approval: https://github.com/canonical/rockcraft/pull/147#pullrequestreview-1192866349 ?

SamirPS commented:
> did something change since the last approval: [#147 (review)](https://github.com/canonical/rockcraft/pull/147#pullrequestreview-1192866349) ?

No, it's was about to know about when it's will be merge but you already answer me

sergiusens commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,122 @@
+How to create a package slice for Chisel
+****************************************
+
+You want to build a chiselled ROCK, but Chisel cannot find the exact bits and pieces of the packages you need to install! This means that said package doesn't yet have the slice definitions you're looking for. In these cases, your best move is to actually **create your own slice definition** (which you can, later on, propose to be merged upstream for everyone else to use).
```
Can we use conditional imperatives here https://diataxis.fr/how-to-guides/#the-language-of-how-to-guides ?

If we move to the imperative part, I would remove "You want to build a chiselled ROCK, but Chisel cannot find the exact bits and pieces of the packages you need to install!"

Side comment, can we standarize on not relying on line wrapping? It makes it easier to review and spatially read the source

"which you can, later on, propose to be merged upstream for everyone else to use" should eventually point to a howto (or where upstream is)

SamirPS replied:
I have modified the ref publish-slice; it refers to the publish slice PR. Is it good now?

SamirPS replied:
@sergiusens the PR looks good to you, or does it need other changes?

sergiusens commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,122 @@
+
+**Let's assume you are trying to create a slice definition for installing the OpenSSL binary into your ROCK!**
+
+Make sure the slice definition doesn't exist yet
```
we should not use contractions https://docs.ubuntu.com/styleguide/en

sergiusens replied:
well, doesn't is allowed....

SamirPS replied:
Can i resolve this conversation?

sergiusens commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,122 @@
+
+Make sure the slice definition doesn't exist yet
+-------------------------------------------------------------
+You obviously want to avoid re-doing existing work, so before even considering creating your own slice definition, please make sure there isn't already something fitting your needs. **How** ?
```
We should not assume something is obvious, switch to the imperative form, we should not be saying please

```suggestion
To avoid re-creating a slice that already exist, check the following to see if something that fits your needs already exists:
```

sergiusens commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,120 @@
+Find the dependencies of your package
+-------------------------------------
+
+So, the first step is to find the dependencies of the package for which you want to create a new slice definitions (``openssl`` in this guide). You can easily do it by:
```
```suggestion
The first step is to find the dependencies of the package for which you want to create a new slice definitions (``openssl`` in this guide). You can easily do it by:
```

tigarmo commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,120 @@
+Make sure the slice definition doesn't exist yet
+------------------------------------------------
+
+To avoid re-creating a slice that already exist, check the following to see if something that fits your needs already exists:
```
```suggestion
To avoid re-creating a slice, check the following to see if something that fits your needs already exists:
```
This way we avoid repeating "exists"

cjdcordeiro replied:
```suggestion
To avoid re-creating a slice that already exists, check the following to see if something that fits your needs already exists:
```

cjdcordeiro commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,120 @@
+
+#. Look into the upstream `chisel-releases <https://github.com/canonical/chisel-releases/tree/ubuntu-22.04/slices>`_ repository
+#. Switch to the branch corresponding to the desired Ubuntu release for your ROCK
+#. Search your package name through the list of slice definition files
```
```suggestion
#. Search your package name within the list of slice definitions files
```

tigarmo commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,120 @@
+Find the dependencies of your package
+-------------------------------------
+
+The first step is to find the dependencies of the package for which you want to create a new slice definitions (``openssl`` in this guide). It's done like this:
```
```suggestion
The first step is to find the dependencies of the package for which you want to create a new slice definition (``openssl`` in this guide). It's done like this:
```

tigarmo commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,120 @@
+Create your slice definition
+----------------------------
+
+You now have everything needed to actually define the OpenSSL slice that will install the content you are looking to have in the final ROCK.Since you are looking to install just the OpenSSL binaries from the ``openssl`` package, what about naming this new slice **bins**? Let's go for it:
```
```suggestion
You now have everything needed to actually define the OpenSSL slice that will install the content you are looking to have in the final ROCK. Since you are looking to install just the OpenSSL binaries from the ``openssl`` package, what about naming this new slice **bins**? Let's go for it:
```

tigarmo commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,120 @@
+#. **What contents do you need from the OpenSSL package?** Just the binaries - ``/usr/bin/c_rehash`` and ``/usr/bin/openssl``
+#. **Does your slice depend on any other package slice?** Yes, OpenSSL depends on ``libc6`` and ``libssl3``
+
+   * **Do these two packages have slice definitions files upstream?** Yes, there is already a slice definitions file for `libc6 <https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libc6.yaml>`_ and another one for `libssl3 <https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libssl3.yaml>`_ . If these dependencies were not present in the upstream Chisel releas ...
```
```suggestion
   * **Do these two packages have slice definitions files upstream?** Yes, there is already a slice definitions file for `libc6 <https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libc6.yaml>`_ and another one for `libssl3 <https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libssl3.yaml>`_ . If these dependencies were not present in the upstream Chisel release, you would also need to create their corresponding slice definitions
```

tigarmo commented:
I started reviewing the text but then more commits and review comments started coming in
I'll finish reviewing after you two go away for the day 😂

cjdcordeiro commented:
> I started reviewing the text but then more commits and review comments started coming in I'll finish reviewing after you two go away for the day joy

my review is done

SamirPS commented:
> I started reviewing the text but then more commits and review comments started coming in I'll finish reviewing after you two go away for the day joy

Waiting for your review

tigarmo commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,120 @@
+Structure of a slice definitions file
+------------------------------------
+
+There is **one slice definitions file** for each Ubuntu package. All of the slice definitions files follow the same structure:
```
```suggestion
There can be only **one slice definitions file** for each Ubuntu package. All of the slice definitions files follow the same structure:
```
My reasoning here is that saying "There is one file for each package" can make people think that every package already has a slice file, which is not true. I believe the intent was to express that each package only has one slice file, right?

SamirPS replied:
For me you are right, i have updated it

cjdcordeiro replied:
:+1:

tigarmo commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,120 @@
+#. **Does your slice depend on any other package slice?** Yes, OpenSSL depends on ``libc6`` and ``libssl3``
+
+   * **Do these two packages have slice definitions files upstream?** Yes, there is already a slice definitions file for `libc6 <https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libc6.yaml>`_ and another one for `libssl3 <https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libssl3.yaml>`_ . If these dependencies were not present in the upstream Chisel releas ...
+   * **Which slices do you depend on then?** Since you only want the OpenSSL binaries, you might only need the libraries from ``libc6`` and ``libssl3``, as well as the configuration files from ``libc6`` and ``openssl`` itself.
```
```suggestion
   * **Which slices do you depend on then?** Since you only want the OpenSSL binaries, you might only need the libraries from ``libc6`` and ``libssl3``, as well as the configuration files from ``libc6`` and ``openssl`` themselves.
```

tigarmo commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,120 @@
+                /usr/lib/ssl/openssl.cnf:
+                /usr/lib/ssl/private:
+
+Notice the unforeseen new slice ``config``. Because your OpenSSL binaries depend on the OpenSSL configuration files, and those were not yet present anywhere in the Chisel releases upstream, you also need to create that slice! You may also ask **"why not put those configuration files inside the ``bins`` slice"**? You could! But we recommend, as a best practice, to separate and group contents accor ...
```
I think something that we need to address in documentation somehow in the future (not on this PR necessarily) is that the jump from "I just need the bins" to "I also need this bunch of config" feels a bit jarring. I know that someone (was it you @cjdcordeiro?) had to "suffer" when creating these slices, but on this how-to the new requirement feels a bit like it came out of nowhere.

In practice, that's probably going to be one of the harder parts when creating new slices, right? Is that something that would merit a separate how-to/troubleshooting/whatever section?

cjdcordeiro replied:
This is really open to developer discretion. Like for DEBs, how do you say what pkgs your DEB depends on? First, the definition must exist, for one to be able to tell what does the slice depends on. In this case, that's the definition we are working on here...it does require some in-depth knowledge about the OpenSSL inner workings

tigarmo commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,120 @@
+   * if you find it, open it and try to find a slice name containing the bits and pieces you need from that package
+
+
+Structure of a slice definitions file
```
This section feels a bit disconnected from the rest of the guide: it reads more like a reference. Is it indispensable here?

cjdcordeiro replied:
could we keep this one here, for now, and add an issue for me to move it to the Chisel reference? (I can't move it now, because we don't have a Chisel reference, and I don't think it should be in the Rockcraft documentation)

cjdcordeiro replied:
created https://github.com/canonical/rockcraft/issues/157

tigarmo commented on docs/how-to/create-slice.rst:
```diff
@@ -0,0 +1,120 @@
+            mutate: |
+                ...
+
+Find the dependencies of your package
```
I like this section, but to me one of the hardest parts of Diátaxis is trying to avoid explaining to much; for how-tos one [should not explain concepts](https://diataxis.fr/how-to-guides/#don-t-explain-concepts) if possible so I actually don't know if this is too much explanation or not! Thoughts?
Maybe we should ask someone on the Documentation team for help? Can either of you attend the Documentation Office Hours today?

SamirPS replied:
When are the Documentation office hours ?

tigarmo replied:
An hour from this comment; this is the meeting: https://meet.google.com/vpj-sanz-mrb

tigarmo replied:
I asked because I can't make it to today's; on second thought it might be better if we finish reviewing this batch and get it merged, consolidate our questions and bring them to Doc office hours next week

cjdcordeiro replied:
IMO, this specific section is fine. It doesn't explain stuff, but rather guide you through a process. The one above though, with the slice schema, indeed feels out of place here. I'll comment there

tigarmo reviewed (approved):
thanks for accommodating all the requested changes!
